Render progress

nbuonin opened the pull request.

sdreher commented on smart_sa/intervention/models.py:
```diff
@@ -869,6 +871,226 @@ def load_game_vars(self, data):
+
+    def session_timestamps(self, session):
+        sv_set = self.sessionvisit_set.all()
+        if not sv_set:
```
A more efficient way to check for an empty queryset is [exists](https://docs.djangoproject.com/en/1.11/ref/models/querysets/#exists).

nbuonin replied:
Ok, got it.

sdreher commented on smart_sa/intervention/models.py:
```diff
@@ -869,6 +871,226 @@ def load_game_vars(self, data):
+        sv_set = self.sessionvisit_set.all()
+        if not sv_set:
+            return []
+        return [sv.logged for sv in sv_set
```
Could this function (and a few similar functions) be simplified by just filtering on the passed session.id and returning the queryset?

sdreher commented on smart_sa/intervention/models.py:
```diff
@@ -869,6 +871,226 @@ def load_game_vars(self, data):
+
+    def mood_score(self):
+        mood_score = self.mood_alcohol_drug_scores()[0]
+        return mood_score if mood_score else None
```
Is 0 a valid mood_score?

nbuonin replied:
Yes it is, I'll simplify this.

sdreher commented on smart_sa/intervention/models.py:
```diff
@@ -936,11 +936,10 @@ def relevant_timestamps(self, session):
     def session_timestamps(self, session):
-        sv_set = self.sessionvisit_set.all()
-        if not sv_set:
+        sv_set = self.sessionvisit_set.all().filter(session_id=session.id)
```
Can remove the `.all()` here, like so: `sv_set = self.sessionvisit_set.filter(session_id=session.id)`

nbuonin replied:
Got it. I fixed it.  I just amended the commit and force pushed, in case you have it checked out locally.

nbuonin commented:
This pull request creates a new report for Masivukeni users. In short it uses a lot of pre-extant code to achieve this.

**Why?**
Masivukeni is being moved to a stateless model.  The end users requested a way to see a users progress on a single page. This was done so that practitioners could see at a glance which parts of the intervention a patient traversed and the scores that they received from certain interactives.

**How?**
After much casting about, I re-used a large amount of code from a dashboard app that had been removed from the site two years ago.  This code provided a number of getter functions on the Participant model that I could call in the templates.  It took me a lot of time to really figure this out because it was unclear to me how instance variables were being created.  Part of this misunderstanding stemmed from that I was unfamiliar with the pattern used to create a new object using class method.  Rather than using init, which is best to not override, its uses a class method to instantiate a Participant object from a JSON string.

Once I wrapped my head around how different attributes of the object were being created, and had copied over the code I needed from the defunct app, I needed to corral these methods into provided the data I needed.  Here I ran into difficulty because I was trying to do too much in the template and not using the view help me.  I got stuck several times trying to iterate of various data structures, which I later realized I could easily manipulate in my view, and then provide any number of variables to my template context.

Once the getters were corralled and I had the data that I needed in my template, I turned to tests.  The existing participant did not have tests associated with it, but the Dashboard app did.  I copied these over, and with much assistance, I modified these to match the getters that I needed.

Lastly I added waffle flags to my templates, to hide all the changes I've made until I'm done polishing.  This was a missed opportunity, I should have added the flags first so that I could incrementally build the page out.
